# Machine Learning - Practical work 01

## 3. Quick numpy introduction

### Max of each row and max of each column

In [28]:
import numpy as np

array1 = np.array([
    [1, 2, 3],
    [4, 5, 6]
])

print("The max of each row is {}".format(np.amax(array1, axis = 1)))
print("The max of each column is {}".format(np.amax(array1, axis = 0)))

The max of each row is [3 6]
The max of each column is [4 5 6]


### Print the elements of array1 that are less than 4

In [31]:
import numpy as np

array1 = np.array([
    [1, 2, 3],
    [4, 5, 6]
])

mask = array1 < 4
print("Elements of array1 that are < 4: {}".format(array1[mask]))

Elements of array1 that are < 4: [1 2 3]


### Convert the values of array2 to the range [0, 1]

In [34]:
array2 = np.array([5, 8, 2, 9, 4, 3])

max = np.amax(array2)
ranger = np.vectorize(lambda t : t/max)
ranger(array2)

array([0.55555556, 0.88888889, 0.22222222, 1.        , 0.44444444,
       0.33333333])

## 4. Loading a dataset with numpy

## 5. Quick Introduction to Pandas

### Méthode choisie pour répondre au problème

- Nous avons utilisé la méthode groupby de DataFrame pour groupé les données en fonctiond de la colonne 'class'.
- Nous n'affichons ensuite que les valeurs liées à la variable 'alcohol'

In [97]:
import pandas as pd
import numpy as np

data = np.genfromtxt('data/wine/wine.data', delimiter=',')
colnames = np.array([
    'class', 'alcohol', 'malic acid', 'ash', 'alcalinity of ash', 'magnesium', 'total phenols', 
    'flavanoids', 'nonflavanoid phenols', 'proanthocyanins', 'color intensity', 'hue',
    'OD280/OD315 of diluted wines', 'proline'
])

df = pd.DataFrame(data=data[:,1:], columns=colnames[1:])
df['class'] = data[:,0].astype(np.int)

print(df.groupby('class').describe()['alcohol'])

       count       mean       std    min     25%     50%     75%    max
class                                                                  
1       59.0  13.744746  0.462125  12.85  13.400  13.750  14.100  14.83
2       71.0  12.278732  0.537964  11.03  11.915  12.290  12.515  13.86
3       48.0  13.153750  0.530241  12.20  12.805  13.165  13.505  14.34


### Résultats

On remarque donc différentes propriétés statistiques des différentes classes de vin. Le tableau ci-dessous résume les informations les plus importantes.

| Classe        | % d'alcool en moyenne | Médiane % d'alcool| Max % d'alcool | Min % d'alcool | 
| ------------- | ------------- | ----- | --- | --- | --- |
| 1             | 13.74 | 13.75 | 14.83 | 12.85 |
| 2             | 12.28 | 12.29 | 13.86 | 11.03 |
| 3             | 13.15 | 13.17 | 14.34 | 12.20 |

## 6. Visual exploratory analysis of data 

### <p style="background-color:#660066; color:#fff;padding:5px; font-weight:bold">Q: By looking at the boxplot, which features seems the most discriminative ?</p>
Il est difficile de désigner un trait de vin pour lequel les classes sont absolument distinctes. Une boîte à moustache permet de visualiser, pour une distribution de valeurs donnée, si une valeur peut être considérée comme aberrante ou non (respectivement en dehors ou en dedans des deux moustaches). Le cas idéal que l'on rechercherait serait donc un trait de vin pour lequel les moustaches des boîtes à moustaches de chaque classe forment des plages qui ne se chevauchent pas. 

Il est évident que nous n'avons pas trouvé un tel trait, nous allons donc considérer les traits dont les boîtes de la boîte à moustache ne se chevauchent pas (ou peu).

On peut donc retenir par exemple les traits de vin suivants comme relativement distincts entre les différentes classes : 
- flavanoids
- proline
- color intensity

## 7. Classifying the wine data

In [2]:
import pandas as pd
import numpy as np

data = np.genfromtxt('data/wine/wine.data', delimiter=',')
colnames = np.array([
    'class', 'alcohol', 'malic acid', 'ash', 'alcalinity of ash', 'magnesium', 'total phenols', 
    'flavanoids', 'nonflavanoid phenols', 'proanthocyanins', 'color intensity', 'hue',
    'OD280/OD315 of diluted wines', 'proline'
])

df = pd.DataFrame(data=data[:,1:], columns=colnames[1:])
df['class'] = data[:,0].astype(np.int)

pred = []

for row in df['alcohol']:
    if row > 13.5:
        pred.append(1);
    elif row > 12.75 and row < 13.5 :
        pred.append(2);    
    else:
        pred.append(3)

# A new column is added to the dataframe
df['prediction'] = pred

len(df.loc[df['class'] == df['prediction']]) / len(df)



0.33707865168539325